hsm/m0hsm.c

```diff
@@ -0,0 +1,748 @@
+		case OPT_MOVE:
+			*flags |= HSM_MOVE;
+			break;
+
+		case OPT_KEEP_PREV_VERS:
+			*flags |= HSM_KEEP_OLD_VERS;
+			break;
+
+		case OPT_WRITE_TO_DEST:
+			*flags |= HSM_WRITE_TO_DEST;
+			break;
+                default:
+                        fprintf(stderr, "Unexpected option: %s\n", value);
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+static int parse_release_subopt(char *opts, enum hsm_rls_flags *flags)
+{
+	enum release_opt {
+		OPT_KEEP_LATEST = 0,
+	};
+	char *const options[] = {
+		[OPT_KEEP_LATEST] = "keep_latest",
+		NULL,
+	};
+	char *subopts = opts;
+	char *value;
+
+	while (*subopts != '\0') {
+		switch (getsubopt(&subopts, options, &value)) {
+		case OPT_KEEP_LATEST:
+			*flags |= HSM_KEEP_LATEST;
+			break;
+
+                default:
+                        fprintf(stderr, "Unexpected option: %s\n", value);
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+static const struct option option_tab[] = {
+	{"quiet", no_argument, NULL, 'q'},
+	{"verbose", required_argument, NULL, 'v'},
+};
+#define SHORT_OPT "qv"
+
+static int parse_cmd_options(int argc, char **argv)
+{
+	int c, option_index = 0;
+
+	while ((c = getopt_long(argc, argv, SHORT_OPT, option_tab,
+				&option_index)) != -1) {
+		switch (c) {
+		case 'q':
+			if (hsm_options.trace_level > LOG_NONE)
+				hsm_options.trace_level--;
+			break;
+		case 'v':
+			if (hsm_options.trace_level < LOG_DEBUG)
+				hsm_options.trace_level++;
+			break;
+		case ':':
+		case '?':
+		default:
+			return -EINVAL;
+		}
+	}
+	return 0;
+}
+
+/* test functions hidden in m0hsm_api */
+int m0hsm_test_write(struct m0_uint128 id, off_t offset, size_t len, int seed);
+int m0hsm_test_read(struct m0_uint128 id, off_t offset, size_t len);
+
+static int run_cmd(int argc, char **argv)
+{
+	struct m0_uint128 id;
+	const char *action;
+	int rc = 0;
+
+	if (!strcmp(argv[optind], "help")) {
+		usage();
+		return 0;
+	}
+
+	/* expect at least cmd <action> <fid> */
+	if (optind > argc - 2) {
+		usage();
+		return -1;
+	}
+
+	action = argv[optind];
+
+	optind++;
+	id = M0_ID_APP;
+	rc = read_fid(argv[optind], &id);
+	if (rc <= 0) {
+		usage();
+		return -1;
+	}
+	optind++;
+
+	if (!strcmp(action, "create")) {
+		int tier;
+		struct m0_obj obj;
+
+		if (optind > argc - 1) {
+			usage();
+			return -1;
+		}
+		tier = atoi(argv[optind]);
+		optind++;
+
+		M0_SET0(&obj);
+		/* Create the object */
+		rc = m0hsm_create(id, &obj, tier, false);
+
+	} else if (!strcmp(action, "show")) {
+		/* get and display the composite layout */
+		rc = m0hsm_dump(stdout, id, false);
+	} else if (!strcmp(action, "dump")) {
+		/* get and display the composite layout */
+		rc = m0hsm_dump(stdout, id, true);
+	} else if (!strcmp(action, "write")) {
+		off_t offset;
+		size_t len;
+		int seed;
+
+		if (optind > argc - 3) {
+			usage();
+			return -1;
+		}
+		offset = read_arg64(argv[optind++]);
+		len = read_arg64(argv[optind++]);
+		seed = atoi(argv[optind]);
+
+		rc = m0hsm_test_write(id, offset, len, seed);
```

Comment:
Now, I understand, 
"write" is a test to write some data (generated with seed).